### ISSUE TITLE
Securing Web API through IdSrv acting as a federation provider

### ISSUE THREAD

**Issue by ajvdburg** [IdentityServer/IdentityServer3#1824]
Hi,

I'm struggling to find the correct configuration for the following scenario: a mobile app uses it's own identity provider which is trusted by my IdSrv instance. It also uses my SDK to connect to my Web API to make certain calls. I obviously need to register the external identity provider with the WS-Fed plugin. But how can I get an access token from my IdSrv to allow the call to the Web API, presuming I secure the API with bearer token authentication?

**Comment by brockallen** [comment]
I'm sort of lost, so nothing is obvious. Can you rephrase please?

Are you asking how do you secure your web api? It needs to trust tokens from IdSvr -- use our AccessTokenValidationMiddleware.

**Comment by ajvdburg** [comment]
Hi,

Sorry, I'll be more specific. Yes I understand that the Web API needs to trust the tokens from the IdSrv with the access token validation middleware. But how does this work with external identity providers? Say I request an access token from an external IdP, can I use that token to authenticate my clients with the Web API using the IdSrv? Or do I need to implement some additional stuff?

**Comment by brockallen** [comment]
If the user is using a browser to authenticate, then it all just works. You're requesting an access token from IdSvr, but IdSvr is using an external IdP to authenticate the user.

**Comment by vijay2358** [comment]
I am not sure this will work external user service or idp or sso logins.. no were given sample and examples.. it is very very difficult to this.. Plz send sample for us.. i am spending lot days to sortout..

**Comment by ajvdburg** [comment]
HI Brock,

Thanks for the reply, using federation with a client browser is indeed not a problem, but say I only have an access/id token from the external IdP? That's the part I'm really struggling to find a solution. Also is mapping the external user to an IdSrv user a must?

**Comment by leastprivilege** [comment]
This is what the assertion flow is about - or custom grant types. Check the docs.

**Comment by ajvdburg** [comment]
Hi,

Checked the docs, and if I understand correctly I could use a custom grant containing the initial access token from the external provider, verify it with that provider to create my own set of claims? Sounds to me like I'm skipping the WS-Federation setup with this implementation, is that a correct assumption?

**Comment by leastprivilege** [comment]
WS-Federation is a protocol for Web SSO - since you are doing API access (and federation) here you are using OAuth 2.0. So yes.

**Comment by ajvdburg** [comment]
Hi,

Ok thanks a lot for all the support! It's been really helpful.
